In [1]:
import numpy as np
import math
import itertools

Problem 33

The fraction $49/98$ is a curious fraction, as an inexperienced mathematician in attempting to simplify it may incorrectly believe that $49/98 = 4/8$, which is correct, is obtained by cancelling the $9$s.
We shall consider fractions like, $30/50 = 3/5$, to be trivial examples.
There are exactly four non-trivial examples of this type of fraction, less than one in value, and containing two digits in the numerator and denominator.
If the product of these four fractions is given in its lowest common terms, find the value of the denominator.


Więc tak: sprawdzam liczby od 10 do 99, jeżeli ich iloraz jest taki jak iloraz ze skreślonymi powtarzającymi się cyframi, to dodaję do list licznik i mianownik. 
Najpierw biorę na tape sytuację, gdzie cyfra jedności licznika skraca cyfrę dziesiątek mianownika. 

In [2]:
nums = []
dens = []
for n in range(10, 99):
    for d in range(n+1, 100):
        n1 = str(n)[0]
        n2 = str(n)[1]
        d1 = str(d)[0]
        d2 = str(d)[1]
        if n2 == d1 and d2 != '0' and n/d == int(n1)/int(d2):
            print(f'{n}/{d}')
            nums.append(n1)
            dens.append(d2)
print([int(n) for n in nums])
print([int(d) for d in dens])

16/64
19/95
26/65
49/98
[1, 1, 2, 4]
[4, 5, 5, 8]


odp: 100

Problem 37

The number $3797$ has an interesting property. Being prime itself, it is possible to continuously remove digits from left to right, and remain prime at each stage: $3797$, $797$, $97$, and $7$. Similarly we can work from right to left: $3797$, $379$, $37$, and $3$.
Find the sum of the only eleven primes that are both truncatable from left to right and right to left.

NOTE: $2$, $3$, $5$, and $7$ are not considered to be truncatable primes.


In [3]:
def prime(n):
    if n < 2: return False
    for k in range(2, int(n**(0.5))+1):
        if n % k == 0:
            return False
    return True

def problem_37():
    liczby = []
    k = 10
    while(len(liczby) < 11):
        bool = True
        if prime(k) == False:
            bool = False
        else:
            for i in range(1, len(str(k))):
                if prime(int(str(k)[i:])) == False or prime(int(str(k)[:-i])) == False:
                    bool = False
                    break

        if bool == True:
            liczby.append(k)
        k+=1

    return(sum(liczby))

Problem 39

If $p$ is the perimeter of a right angle triangle with integral length sides, $\{a, b, c\}$, there are exactly three solutions for $p = 120$.
$\{20,48,52\}$, $\{24,45,51\}$, $\{30,40,50\}$
For which value of $p \le 1000$, is the number of solutions maximised?


Muszę określić granice długości dla przeciwprostokątnej: nie może być dłuższa niż podłoga połowy obwodu. Nie może być krótsza niż (podłoga ?) z obwodu razy ((2**0.5)/(2 + 2**0.5) = 0.4142135623730951). Przechowuję liczbę możliwych trójkątów i obwód, zmieniam kiedy trzeba. 

prostokątny równoramienny o długościach 4, 4, 5.656854249492381

In [4]:
const = 0.4142135623730951
p_max = 0
triangles_max = 0
for p in range(10, 1001):
    curr_triangles = 0
    for c in range(int(p*const), int(p/2)+1):
        for b in range(math.ceil((p-c)/2), p - c):
            if (p-c-b)**2 + b**2 == c**2:
                curr_triangles += 1
                print(f'obwód: {p}, boki: {[c, b, p-c-b]}')
    if curr_triangles > triangles_max:
        p_max = p
        triangles_max = curr_triangles
print([p_max, triangles_max])

obwód: 12, boki: [5, 4, 3]
obwód: 24, boki: [10, 8, 6]
obwód: 30, boki: [13, 12, 5]
obwód: 36, boki: [15, 12, 9]
obwód: 40, boki: [17, 15, 8]
obwód: 48, boki: [20, 16, 12]
obwód: 56, boki: [25, 24, 7]
obwód: 60, boki: [25, 20, 15]
obwód: 60, boki: [26, 24, 10]
obwód: 70, boki: [29, 21, 20]
obwód: 72, boki: [30, 24, 18]
obwód: 80, boki: [34, 30, 16]
obwód: 84, boki: [35, 28, 21]
obwód: 84, boki: [37, 35, 12]
obwód: 90, boki: [39, 36, 15]
obwód: 90, boki: [41, 40, 9]
obwód: 96, boki: [40, 32, 24]
obwód: 108, boki: [45, 36, 27]
obwód: 112, boki: [50, 48, 14]
obwód: 120, boki: [50, 40, 30]
obwód: 120, boki: [51, 45, 24]
obwód: 120, boki: [52, 48, 20]
obwód: 126, boki: [53, 45, 28]
obwód: 132, boki: [55, 44, 33]
obwód: 132, boki: [61, 60, 11]
obwód: 140, boki: [58, 42, 40]
obwód: 144, boki: [60, 48, 36]
obwód: 144, boki: [65, 63, 16]
obwód: 150, boki: [65, 60, 25]
obwód: 154, boki: [65, 56, 33]
obwód: 156, boki: [65, 52, 39]
obwód: 160, boki: [68, 60, 32]
obwód: 168, boki: [70, 56, 42]
obwó

Problem 45 - ZROBIONY

Triangle, pentagonal, and hexagonal numbers are generated by the following formulae:

Triangle: $T_n=n(n+1)/2$
 
$1, 3, 6, 10, 15, \dots$

Pentagonal: 
$P_n=n(3n - 1)/2$
 
$1, 5, 12, 22, 35, \dots$

Hexagonal: 
$H_n=n(2n - 1)$
 
$1, 6, 15, 28, 45, \dots$
It can be verified that $T_{285} = P_{165} = H_{143} = 40755$.
Find the next triangle number that is also pentagonal and hexagonal.


Ok, to mam może głupi pomysł, ale trzymać trzy listy zawierające liczby triagonalne, pentagonalne i heksagonalne, i dodaję następną liczbę do listy, której ostatnia wartość w liście jest najmniejsza. 
Czyli T=[1, 3], P=[1, 5], H=[1, 6], dodaję następną do T i mam T=[1, 3, 6], dodaję 12 do P i tak dalej, za każdym razem sprawdzając czy ostatnie pozycje się zgadzają. 
Czyli każda iteracja to jest:
- sprawdzenie równości ostatnich wartości w tabelach, jeśli jest równość to koniec (pod warunkiem że ta wartość to nie 40755)
- jeśli nie ma równości, to do tablicy o najmniejszej ostatniej wartości dodaję kolejny element
PYTANIE: jak na podstawie ostatniego elementu dostać tablicę?
mam tylko 3 tablice, więc mogę zrobić to po chamsku if if if, trochę szkoda by było to tak zrobić
Najpierw zrobię po chamsku a potem spytam wujka google o pomoc

In [5]:
# def zwieksz_mniejsza(t1, t2):
#     if t1[-1]<t2[-1]:
#         t1.append(t1[-1]+1)
#         print('dodałem do pierwszej')
#     elif t1[-1]>t2[-1]:
#         t2.append(t2[-1]+1)
#         print('dodałem do drugiej')
#     else:
#         print('wartości ostatnich elementów w obu tablicach są równe')

# a = [123,345,675]
# b = [345,5687]
# zwieksz_mniejsza(a, b)
# print(f'{a}\n{b}')
# print(sorted(a, reverse=True))

Może mógłbym tworzyć słownik nazwy tablicy i wartości ostatniego elementu? Tylko kluczem musiałby być string, jak potem stringiem powołać się na tablicę jeśli nie przez ify?

In [6]:
tn = lambda n: n*(n-1)/2
pn = lambda n: n*(3*n-1)/2
hn = lambda n: n*(2*n-1)


t = [1, 3]
p = [1, 5]
h = [1, 6]
while True:
    if t[-1] == p[-1] and p[-1] == h[-1] and t[-1] != 40755:
        print(f'Nasza liczba to {int(t[-1])}')
        break
    sorted_tables = sorted([t[-1], p[-1], h[-1]])
    if sorted_tables[0] == t[-1]:
        t.append(tn(len(t) + 1))
    if sorted_tables[0] == p[-1]:
        p.append(pn(len(p) + 1))
    if sorted_tables[0] == h[-1]:
        h.append(hn(len(h) + 1))
    

Nasza liczba to 1533776805


Problem 27 ZROBIONY

Euler discovered the remarkable quadratic formula:

$$n^2 + n + 41$$

It turns out that the formula will produce **40 primes** for consecutive integer values $0 \le n \le 39$.

However, when $n = 40$:

$$40^2 + 40 + 41 = 40(40 + 1) + 41$$

which is divisible by $41$.

And when $n = 41$:

$$41^2 + 41 + 41$$

is clearly divisible by $41$.

---

The incredible formula:

$$n^2 - 79n + 1601$$

produces **80 primes** for consecutive values $0 \le n \le 79$.

The product of the coefficients $-79$ and $1601$ is:

$$-126479$$

---

Considering quadratics of the form:

> $$n^2 + an + b$$
> where $|a| < 1000$ and $|b| \le 1000$
>
> where $|n|$ is the absolute value of $n$
> e.g. $|11| = 11$ and $|-4| = 4$

---

**Task:**
Find the product of the coefficients $a$ and $b$ for the quadratic expression that produces the maximum number of primes for consecutive values of $n$, starting with $n = 0$.


Wcześniejsza funkcja prime() szybko rozpoznaje liczby ujemne, więc dobrze. 
robię pętlę w pętli (4 mln wykonań). Potrzebuję 3 zmiennych tymczasowych (na aktualne współczynniki i długość ciągu liczb pierwszych) i 3 zmiennych stałych (na najlepsze współczynniki i długość ciągu liczb pierwszych). W każdym wykonaniu pętli wrzucam kolejne liczby naturalne aż któraś wartość nie będzie pierwsza. Jeśli aktualna długość ciągu jest większa od najdłuższego ciągu, to zmieniam, na końcu programu zwracam iloczyn a*b. 

In [7]:
best_a = -999
best_b = -999
longest_seq = 0

for a in range(-999, 1000):
    for b in range(-999, 1000):
        seq = 0 
        v = lambda n: int(n**2 + a*n + b)
        n = 0
        while True:
            if prime(v(n)) == False:
                break
            else:
                seq += 1 # jeśli wartość dla tego n jest pierwsza to zwiększ długość aktualnego ciągu
                n += 1
        if seq > longest_seq:
            best_a = a
            best_b = b
            longest_seq = seq
print([best_a, best_b, longest_seq])
print(f'Odpowiedź: {best_a*best_b}')

[-61, 971, 71]
Odpowiedź: -59231


Problem 35 ZROBIONY

The number $197$ is called a **circular prime** because all rotations of its digits — $197$, $971$, and $719$ — are themselves prime.

There are **thirteen** such primes below $100$:

$$
2, 3, 5, 7, 11, 13, 17, 31, 37, 71, 73, 79, 97
$$

---

**Task:**
How many circular primes are there below $1,000,000$?


Pierwsza rzecz: wśród cyfr nie może być żadnej parzystej, bo w pewnej permmutacji parzysta jest na końcu i ta permutacja na pewno nie jest pierwsza. Z tego samego powodu wśród cyfr nie może być 5. UWAGA: 2 i 5 są pierwsze, więc przy powyższym filtrowaniu trzeba je dodać osobno. 
Kolejny problem to sprawdzenie wszystkich permutacji. Może jest gotowa funkcja? Jeśli nie, to trzeba skonstruować algorytm generowania permutacji (w pierwszym pliku jest gdzieś w porządku leksykograficznym). UWAGA: cyfry mogą się powtarzać!! WNIOSEK: lepszy byłby algorytm iteracyjny, nie rekurencyjny (bo powyższa uwaga + szybszy)

Może jednak jakiś rekurencyjny na zasadzie "permutacja n elementowego to zabranie jakiegoś elementu i wsadzenie na każde miejsce permutacji pozostałych"

mogę użyć metody itertools.permutations()

In [8]:
def parzyste_cyfry(n): # zakładam że przyjmuje int
    n = str(n)
    for i in n: # tutaj i to str!
        if int(i) % 2 == 0 or i == '5':
            return True
    return False
print([parzyste_cyfry(k) for k in [2, 3, 11, 27, 747]])

[True, False, False, True, True]


In [9]:
pierwsze_nieparzyste_cyfry = [2, 3, 5] + [p for p in range(6, 10**6) if prime(p) and parzyste_cyfry(p) == False]
print(pierwsze_nieparzyste_cyfry)

[2, 3, 5, 7, 11, 13, 17, 19, 31, 37, 71, 73, 79, 97, 113, 131, 137, 139, 173, 179, 191, 193, 197, 199, 311, 313, 317, 331, 337, 373, 379, 397, 719, 733, 739, 773, 797, 911, 919, 937, 971, 977, 991, 997, 1117, 1171, 1193, 1319, 1373, 1399, 1733, 1777, 1913, 1931, 1933, 1973, 1979, 1993, 1997, 1999, 3119, 3137, 3191, 3313, 3319, 3331, 3371, 3373, 3391, 3719, 3733, 3739, 3779, 3793, 3797, 3911, 3917, 3919, 3931, 7177, 7193, 7331, 7333, 7393, 7717, 7793, 7919, 7933, 7937, 7993, 9133, 9137, 9173, 9199, 9311, 9319, 9337, 9371, 9377, 9391, 9397, 9719, 9733, 9739, 9791, 9931, 9973, 11113, 11117, 11119, 11131, 11171, 11173, 11177, 11197, 11311, 11317, 11393, 11399, 11717, 11719, 11731, 11777, 11779, 11933, 11939, 11971, 13171, 13177, 13313, 13331, 13337, 13339, 13397, 13399, 13711, 13799, 13913, 13931, 13933, 13997, 13999, 17117, 17137, 17191, 17317, 17333, 17377, 17393, 17713, 17737, 17791, 17911, 17939, 17971, 17977, 19139, 19319, 19333, 19373, 19379, 19391, 19717, 19739, 19777, 19793, 19913,

Warto zauważyć, że tutaj tablica 'pierwsze_nieparzyste_cyfry' zawiera liczby pierwsze od 2 do 10**6-1 nie licząc tych, które mają cyfry parzyste lub 5 (2 i 5 należy do tej tablicy). 
Czy mogę "zmergować" tupletkę zawierającą cyfry jako znaki?

In [10]:
licznik = 0
def tupl_to_int(tupletka):
    string = ''
    for i in tupletka:
        string = string + i
    return(int(string))

for p in pierwsze_nieparzyste_cyfry:
    bool = True
    for perm in itertools.permutations(str(p)): # perm to tupletka
        if prime(tupl_to_int(perm)) == False:
            bool = False
            break
    if bool == True:
        licznik += 1
        print(p)
print(f'Odpowiedź: {licznik}')

2
3
5
7
11
13
17
31
37
71
73
79
97
113
131
199
311
337
373
733
919
991
Odpowiedź: 22


Źle, nie ma np. 197

Dobra, teraz rozumiem że nie chodzi o permutacje tylko rotacje

In [11]:
def rotacje(k):
    rotacje = []
    for i in range(len(k)):
        k = k[-1] + k[:-1]
        rotacje.append(k)
    return rotacje

In [12]:
licznik = 0

for p in pierwsze_nieparzyste_cyfry:
    bool = True
    for c in rotacje(str(p)): # perm to tupletka
        if prime(int(c)) == False:
            bool = False
            break
    if bool == True:
        licznik += 1
print(f'Odpowiedź: {licznik}')

Odpowiedź: 55


Problem 49

The arithmetic sequence $1487, 4817, 8147$, in which each term increases by $3330$, is unusual in two ways:

1. Each of the three terms is **prime**.
2. Each of the $4$-digit numbers is a **permutation** of the others.

There are no arithmetic sequences made up of three $1$-, $2$-, or $3$-digit primes exhibiting this property, but there is **one other** $4$-digit increasing sequence.

---

**Task:**
What $12$-digit number do you form by concatenating the three terms in this sequence?


Więc tak, od początku:
- Najpierw mógłbym znaleźć wszystkie liczby pierwsze od 1000 do 9999. 
- potem mógłbym je podzielić na klasy abstrakcji, gdzie aRb wtw gdy a jest permutacją b
Tu rodzi się problem, co jeśli w klasie abstrakcji, w której jest Szukana, są cztery liczby i ciąg arytmetyczny tworzą np. pierwsza, trzecia i czwarta? Spróbuję zająć się tym problemem potem, najpierw zobaczę ile jest tych klas abstrakcji. 

In [ ]:
# coś nie hula

# pierwsze_int = [p for p in range(1000, 10000) if prime(p)] # około tysiąc liczb, są posortowane
# pierwsze_tuple = [tuple(str(p)) for p in pierwsze]
# klasy_abstrakcji = []
# while len(pierwsze_tuple) > 0:
#     klasy_abstrakcji.append([pierwsze_tuple]) # tworzę nową klasę abstrakcji
#     for t in pierwsze_tuple:
#         if t in itertools.permutations(klasy_abstrakcji[-1][0]):
#             klasy_abstrakcji[-1].append(t)
#             pierwsze_tuple.remove(t)
# print(klasy_abstrakcji)



0
